src/tv2-common/helpers/graphics/Graphic.ts

```diff
@@ -0,0 +1,171 @@
+import { PieceLifespan, TSR } from 'blueprints-integration'
+import {
+	calculateTime,
+	CueDefinitionGraphic,
+	ExtendedShowStyleContext,
+	getDefaultOut,
+	getLifeSpan,
+	getTimingEnable,
+	GraphicInternalOrPilot,
+	GraphicIsInternal,
+	IsTargetingTLF,
+	IsTargetingWall,
+	TableConfigItemGfxTemplate,
+	TV2ShowStyleConfig
+} from 'tv2-common'
+import { GraphicEngine, SharedSourceLayers } from 'tv2-constants'
+
+export abstract class Graphic {
+	protected readonly config: TV2ShowStyleConfig
+	protected readonly engine: GraphicEngine
+	constructor(
+		protected context: ExtendedShowStyleContext,
+		protected cue: CueDefinitionGraphic<GraphicInternalOrPilot>
+	) {
+		this.config = context.config
+		this.engine = cue.target
+	}
+
+	public abstract getTemplateId(): string
+	public abstract getTemplateName(): string
+
+	protected getGraphicDuration(): number | undefined {
+		if (this.config.showStyle.GfxTemplates) {
+			const template = this.findGfxTemplate()
+			if (template && template.OutType && !template.OutType.toString().match(/default/i)) {
+				return undefined
+			}
+		}
+
+		return getDefaultOut(this.config)
+	}
+
+	protected getSourceLayerForGraphic(name: string) {
+		const conf = this.config.showStyle.GfxTemplates
+			? this.config.showStyle.GfxTemplates.find((gfx) => gfx.VizTemplate.toString() === name)
+			: undefined
+
+		if (!conf) {
+			return SharedSourceLayers.PgmGraphicsOverlay
+		}
+
+		switch (conf.SourceLayer) {
+			// TODO: When adding more sourcelayers
+			// This is here to guard against bad user input
+			case SharedSourceLayers.PgmGraphicsHeadline:
+				if (this.config.studio.GraphicsType === 'HTML') {
+					return SharedSourceLayers.PgmGraphicsLower
+				}
+				return SharedSourceLayers.PgmGraphicsHeadline
+			case SharedSourceLayers.PgmGraphicsIdent:
+				return SharedSourceLayers.PgmGraphicsIdent
+			case SharedSourceLayers.PgmGraphicsLower:
+				return SharedSourceLayers.PgmGraphicsLower
+			case SharedSourceLayers.PgmGraphicsOverlay:
+				return SharedSourceLayers.PgmGraphicsOverlay
+			case SharedSourceLayers.PgmGraphicsTLF:
+				return SharedSourceLayers.PgmGraphicsTLF
+			case SharedSourceLayers.PgmGraphicsTema:
+				return SharedSourceLayers.PgmGraphicsTema
+			case SharedSourceLayers.PgmGraphicsTop:
+				return SharedSourceLayers.PgmGraphicsTop
+			case SharedSourceLayers.WallGraphics:
+				return SharedSourceLayers.WallGraphics
+			default:
+				return SharedSourceLayers.PgmGraphicsOverlay
+		}
+	}
+
+	protected createTimingGraphic(): { start: number; duration?: number } {
+		const ret: { start: number; duration?: number } = { start: 0, duration: 0 }
+		const start = this.cue.start ? calculateTime(this.cue.start) : 0
+		start !== undefined ? (ret.start = start) : (ret.start = 0)
+
+		const duration = this.getGraphicDuration()
+		const end = this.cue.end
+			? this.cue.end.infiniteMode
```

Comment:
Nesting ternaries is a very bad practice as they become very difficult to read very fast

Reply:
Old old code that I thought I already refactored some time ago, but apparently not, so I did now